libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * @module botbuilder-dialogs-adaptive
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Recognizer } from '../recognizers';
+import { StringExpression, ArrayExpression } from 'adaptive-expressions';
+import { LuisPredictionOptions, LuisRecognizerOptionsV3, LuisRecognizer, LuisApplication } from 'botbuilder-ai';
+import { Configurable, DialogContext } from 'botbuilder-dialogs';
+import { BotTelemetryClient, Activity, ActivityTypes } from 'botbuilder-core';
+
+export class LuisAdaptiveRecognizer extends Configurable implements Recognizer {
+    /**
+     * Recognizers unique ID.
+     */
+    public id: string;
+
+    /**
+     * LUIS application ID.
+     */
+    public applicationId: StringExpression;
+
+    /**
+     * LUIS dynamic list.
+     */
+    public dynamicLists: ArrayExpression<any>;
+
+    /**
+     * LUIS endpoint to query.
+     * @remarks
+     * For example: "https://westus.api.cognitive.microsoft.com"
+     */
+    public endpoint: StringExpression;
+
+    /**
+     * Key used to talk to a LUIS endpoint.
+     */
+    public endpointKey: StringExpression;
+
+    /**
+     * External entity recognizer.
+     * @remarks
+     * This recognizer is run before calling LUIS and the results are passed to LUIS.
+     */
+    // public externalEntityRecognizer: Recognizer;
+
+    /**
+     * LUIS prediction options.
+     */
+    public predictionOptions: LuisPredictionOptions;
+
+
+    /**
+     * Telemetry client.
+     */
+    public telemetryClient: BotTelemetryClient;
+
+    public async recognize(dialogContext: DialogContext, activity: Activity) {
+        // Validate passed in activity matches turn activity 
+        const context = dialogContext.context
+        if (activity && (!context.activity == null || context.activity.type != ActivityTypes.Message || context.activity.text != activity.text)) {
```

Comment:
`!context.activity == null` should never evaluate to true.

I've refactored the code a little in an attempt to improve readability and used strict non-equality checks:

```suggestion
        const context = dialogContext.context;
        const utteranceMatches: boolean = (context.activity !== undefined && context.activity !== null)
            || context.activity.type !== ActivityTypes.Message
            || context.activity.text != activity.text;
        
        if (activity && utteranceMatches) {
```

I left `context.activity.text != activity.text`, but maybe that should be changed as well?

Reply:
Good catch... I forgot to remove the "= null" part

Reply:
Your version had issues as well but the changed version should work

---
In reply to: [418658435](https://github.com/microsoft/botbuilder-js/pull/2149#discussion_r418658435) [](ancestors = 418658435)